pkg/policy/controller.go

```diff
@@ -174,9 +179,14 @@ func (pc *PolicyController) updatePolicy(old, cur interface{}) {
 	pc.pMetaStore.UnRegister(*oldP)
 	pc.pMetaStore.Register(*curP)
 
+	if curP.Spec.Background == nil {
+		// skipped policy mutation default -> do not process
+		return
+	}
+
 	// Only process policies that are enabled for "background" execution
 	// policy.spec.background -> "True"
-	if !curP.Spec.Background {
+	if !*curP.Spec.Background {
```

Comment:
Can we also check if the curP has userInfo defined, if there's no userInfo then push it to the queue? With this kyverno would: 1. report violations for the pre-existing policies (with no userInfo); 2. register resource webhook when there's the policy exist in the cluster

Reply:
This is done in policy validation, where we don't allow `Background` flag if user is using userInfo in filters, or variables. An error is returned

Reply:
It would be easier to use only one parameter for background processing or not, instead of multiple checks.

Reply:
But don't the policy validation only performs on new incoming policies? How we verify it for the exsiting one?

Reply:
Yes, but that will be a limitation. As none of the changes in policy webbooks will be applied to a policy that includes all the defaulting and validation.

Reply:
I can add a check for that but then we need to add checks for all changes in webhook checks in the controller.

Reply:
>add checks for all changes in webhook

So we are not pushing to the queue for policies without userInfo, but to report errors? How does the user know the errors for the pre-exist policies?

Reply:
I don't understand the question. To give an overview of the current code:
- Policy Validation: If there is userInfo used with background flag, we report the error to the user. so the policy is never created.
- The policy controller manages polices for background processing. If it has background flag enabled then it will be processed in the background.

We don't reconcile errors due to previous policies, as we rely on the admission webhook.  So we cannot do that unless we duplicate each check-in policy validation & mutation in the background as well, but that's not the flow for kyverno.

What you are suggesting is fair, but it is limitation of using admission controls.
@realshuting let me know if there are more questions on this PR, as most of the question is around the flow of kyverno and can be factored separately.

Reply:
The scenario is what I described above: 1. the user has the policies deployed 2. then the user installs Kyverno. With the flow in PR, what will happen is (assume there's no incoming policy to be created):
- the violations would not be reported for those policies without userInfo defined
- as we don't enqueue any pre-deployed policies, the resource webhook would not be registered by policy controller, this will lead to the result (**BUG**) that we are not applying the existing policies to  the incoming resource due to there's no webhook. We pushed in a temporary fix in #589, so either we keep [this](https://github.com/nirmata/kyverno/blob/622d007e18dbcd743d2211a6a0f364eef86bf7b5/pkg/policy/controller.go#L251) resource webhook register added outside of synchandler (which is not recommended as you explained [here](https://github.com/nirmata/kyverno/pull/589#issuecomment-570453491) ),  or we register the resource webhook somewhere else.

So the suggestion was to enqueue the policies by checking if `curP.Spec.Background==true` or `userInfo={}`. We can have alternative solutions as long as the above case is handled.

Reply:
Again, the concern is fair but will have to apply for all logic in policy validation and policy mutation. (which includes, defaulting, pod-controllers and others). I'll add the temp fix as suggested, and create a separate issue to track the above scenarios. As we need to handle all possible scenarios for all features instead of cherry-picking some.

Reply:
Added the check in commit https://github.com/nirmata/kyverno/pull/594/commits/eadf3fc71794db28bc07451b907657f0ec6cef67